File: tcc.py
import os
import soundfile
import time
from onnxexp import export_tcc
from torch import (
    cuda,
    load,
    no_grad,
)
from model import SpeakerEmbeddingExtractor, ToneColorConverter, VoiceActivityDetector


FILTER_LENGTH = 1024
HOP_LENGTH = 256
WIN_LENGTH = 1024
SAMPLING_RATE = 22050
GIN_CHANNELS = 256

asset_dir = 'asset'
model_dir = "checkpoint"
device = "cuda:0" if cuda.is_available() else "cpu"

tcc = ToneColorConverter(
    FILTER_LENGTH, GIN_CHANNELS, HOP_LENGTH, SAMPLING_RATE, WIN_LENGTH
).to(device)
tcc.load_state_dict(load(os.path.join(model_dir, "tone_color_converter.pt")))
see = SpeakerEmbeddingExtractor(
    FILTER_LENGTH, GIN_CHANNELS, HOP_LENGTH, SAMPLING_RATE, WIN_LENGTH
).to(device)
see.load_state_dict(load(os.path.join(model_dir, "speaker_embedding_extractor.pt")))
vad = VoiceActivityDetector(os.path.join(model_dir, "voice_activity_detector.onnx"))

tgt_se, _ = see.get_se(os.path.join(asset_dir, 'houge.mp3'), vad, device)
src_se, src_audio = see.get_se(os.path.join(asset_dir, 'test7.wav'), vad, device)

with no_grad():
    t = time.time()
    audio = tcc(src_audio, src_se, tgt_se).cpu().numpy()
    print(time.time() - t)
    soundfile.write("out.wav", audio, SAMPLING_RATE)
    export_tcc(
        tcc,
        os.path.join(model_dir, "tone_color_converter.onnx"),
        src_audio,
        src_se,
        tgt_se,
    )

File: voxudio-rs/examples/tcc.rs
//! 语音转换示例程序
//!
//! 该程序演示了如何使用voxudio库进行语音活动检测、说话人嵌入提取和音色转换
//! 主要流程:
//! 1. 加载源音频和目标音频
//! 2. 使用VAD保留语音部分
//! 3. 提取说话人嵌入特征
//! 4. 进行音色转换
//! 5. 播放转换后的音频

use voxudio::*;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let mut vad = VoiceActivityDetector::new("../checkpoint/voice_activity_detector.onnx")?;
    let mut see = SpeakerEmbeddingExtractor::new("../checkpoint/speaker_embedding_extractor.onnx")?;
    let mut tcc = ToneColorConverter::new("../checkpoint/tone_color_converter.onnx")?;

    let (src_audio, src_channels) = load_audio::<22050, _>("../asset/test6.wav", false).await?;
    let vad_audio = vad
        .retain_speech_only::<22050>(&src_audio, src_channels)
        .await?;
    let src_se = see.extract(&vad_audio, src_channels).await?;

    let (tgt_audio, tgt_channels) = load_audio::<22050, _>("../asset/bajie.mp3", false).await?;
    let vad_audio = vad
        .retain_speech_only::<22050>(&tgt_audio, tgt_channels)
        .await?;
    let tgt_se = see.extract(&vad_audio, tgt_channels).await?;

    let (out_audio, out_channels, cost) = tcc.convert(&src_audio, &src_se, &tgt_se).await?;
    println!("Convert cost: {:?}", cost);

    let mut ap = AudioPlayer::new()?;
    ap.play()?;
    ap.write::<22050>(&out_audio, out_channels).await?;

    Ok(())
}